lib/jekyll/data_hash.rb

```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module Jekyll
+  # A class that behaves very similar to Ruby's `Hash` class yet different in how it is handled by
+  # Liquid. This class emulates Hash by delegation instead of inheritance to minimize overridden
+  # methods especially since some Hash methods returns another Hash instance instead of the
+  # subclass instance.
+  class DataHash
+    #
+    # Delegate given (zero-arity) method(s) to the Hash object stored in instance variable
+    # `@registry`.
+    # NOTE: Avoiding the use of `Forwardable` module's `def_delegators` for preventing unnecessary
+    # creation of interim objects on multiple calls.
+    def self.delegate_to_registry(*symbols)
+      symbols.each { |sym| define_method(sym) { @registry.send(sym) } }
+    end
+    private_class_method :delegate_to_registry
+
+    # -- core instance methods --
+
+    attr_accessor :context
+
+    def initialize
+      @registry = {}
+    end
+
+    def [](key)
+      @registry[key].tap do |value|
+        value.context = context if value.respond_to?(:context=)
```

Comment:
Is the only reason we have DataHash for this one line?

Reply:
Primarily yes. But also to avoid manipulating the Ruby Core class `Hash`.
Ultimately, we need a way to detect what entities are dependent on a given data file or directory of data files to trigger the entity's re-rendering when data file(s) change. The most common way end-users directly consume `Jekyll::Site#site_data` is via Liquid drops: `{{ site.data[my_var] }}`. Therefore, the only access to the parent entity's `path` attribute is via the  Liquid `context`. So, we need an object that behaves like Ruby's own `Hash` yet be aware of render context **and** remain the same subclass after execution of certain methods (that return a Hash object irrespective of the caller class).
The last point is debatable because I don't quite remember which methods exactly and therefore cannot assert whether modern Ruby versions have addressed that oddity.

Reply:
Gotcha, thanks for the reiteration.